src/main/java/org/candlepin/subscriptions/tally/admin/MergeHostsMigration.java

```diff
@@ -0,0 +1,240 @@
+    if (primaryHost.getHypervisorUuid() == null)
+      primaryHost.setHypervisorUuid(host.getHypervisorUuid());
+    if (primaryHost.getBillingProvider() == null)
+      primaryHost.setBillingProvider(host.getBillingProvider());
+    if (primaryHost.getBillingAccountId() == null)
+      primaryHost.setBillingAccountId(host.getBillingAccountId());
+    // keep the latest applied event record date
+    if (primaryHost.getLastAppliedEventRecordDate() == null
+        || host.getLastAppliedEventRecordDate() != null
+            && primaryHost
+                    .getLastAppliedEventRecordDate()
+                    .compareTo(host.getLastAppliedEventRecordDate())
+                < 0) {
+      primaryHost.setLastAppliedEventRecordDate(host.getLastAppliedEventRecordDate());
+    }
+    // HBI_HOST will always win if there are different values
+    if (primaryHost.getInstanceType() == null
+        || !primaryHost.getInstanceType().equals("HBI_HOST") && host.getInstanceType() != null) {
+      primaryHost.setInstanceType(host.getInstanceType());
+    }
+  }
+
+  /**
+   * Compile all buckets from all hosts
+   *
+   * @param primaryHost
+   * @param host
+   */
+  private void mergeTallyBuckets(Host primaryHost, Host host) {
+    for (HostTallyBucket hostTallyBucket : host.getBuckets()) {
+      primaryHost.addBucket(
+          hostTallyBucket.getKey().getProductId(),
+          hostTallyBucket.getKey().getSla(),
+          hostTallyBucket.getKey().getUsage(),
+          hostTallyBucket.getKey().getBillingProvider(),
+          hostTallyBucket.getKey().getBillingAccountId(),
+          hostTallyBucket.getKey().getAsHypervisor(),
+          hostTallyBucket.getSockets(),
+          hostTallyBucket.getCores(),
+          hostTallyBucket.getMeasurementType());
+    }
+  }
+
+  /**
+   * Add measurements to primary only if it does not exist yet
+   *
+   * @param primaryHost
+   * @param host
+   */
+  private void mergeMeasurements(Host primaryHost, Host host) {
+    for (String measurementName : host.getMeasurements().keySet()) {
+      if (primaryHost.getMeasurements().get(measurementName) == null) {
+        primaryHost
+            .getMeasurements()
+            .put(measurementName, host.getMeasurements().get(measurementName));
+      }
+    }
+  }
+
+  /**
+   * Add monthly totals to primary; duplicates will be handled in the Host object
+   *
+   * @param primaryHost
+   * @param host
+   */
+  private void mergeMonthlyTotals(Host primaryHost, Host host) {
+    // add monthly totals to primary. duplicates will be handled in the Host object
+    for (InstanceMonthlyTotalKey monthlyKey : host.getMonthlyTotals().keySet()) {
+      primaryHost.addToMonthlyTotal(
+          monthlyKey.getMonth(),
+          MetricId.fromString(monthlyKey.getMetricId()),
+          host.getMonthlyTotal(
+              monthlyKey.getMonth(), MetricId.fromString(monthlyKey.getMetricId())));
+    }
+  }
+
+  @Override
+  public void recordCompleted() {
```

Comment:
I think this indicates that there's some future work to be done around the DataMigration class and what it's used for.  This seems to be the second time we've extended it and then opted to intentionally leave the implementation for recordCompleted() blank.  To me, this indicates maybe we're not using DataMigration as it was originally intended (also an indicator is how there are references to liquibase changesets in DataMigration) for.

For the sake of this story/PR, I'm ok with it as it is because it's working and adding value, but I'd be interested in having a conversation with @kahowell at some point about intended usage of that abstract class.